### ISSUE TITLE
Version-gated tokenizer file selection (fast_tokenizer_files) hands different vocabularies to different consumers of the same repo

### ISSUE THREAD

**Issue by AUTHENSOR** [huggingface/transformers#48836]
# Version-gated tokenizer file selection (fast_tokenizer_files) hands different vocabularies to different consumers of the same repo+revision, silently

**Severity: high** (a model repo author can make a serving stack and a safety checker disagree on what the same token ids say; no warning is emitted and no record of the chosen file exists on the loaded object)

**Affected versions:** transformers 4.57.1 (PyPI, runtime-verified; the installed tokenization_utils_base.py was checked byte-identical to the v4.57.1 tag) and the 5.x dev line at commit `78fa17cb266f566f339d85acb8ecafc675d1119e`; the same selection logic stands in both. As of 2026-09-15 the selection code is unchanged on main.

## Summary

`PreTrainedTokenizerBase.from_pretrained` honors a `fast_tokenizer_files` list in `tokenizer_config.json`, and `get_fast_tokenizer_file` picks the highest-versioned file whose version is at or below the INSTALLED transformers version:

```python
    transformers_version = version.parse(__version__)
    for v in available_versions:
        if version.parse(v) <= transformers_version:
            tokenizer_file = tokenizer_files_map[v]
        else:
            # No point going further since the versions are sorted.
            break
```

(src/transformers/tokenization_utils_base.py, `get_fast_tokenizer_file`, lines 3458-3486 on the 5.x dev line; line 4190 on 4.57.1. Call site: lines 1652-1653 on 5.x dev, lines 2024-2025 on 4.57.1.)

This is an intentional upgrade mechanism: the repo can ship `tokenizer.json` plus `tokenizer.9.9.9.json`, and consumers on transformers >= 9.9.9 get the new vocabulary while older consumers keep the old one. The library's own test suite exercises exactly this split (`TokenizerVersioningTest` against the `hf-internal-testing/test-two-tokenizers` fixture).

The consequence that follows from this plus three ordinary deployment patterns is that "the tokenizer for model X" is not a well-defined object: which vocabulary decodes a given token-id list depends on the consumer, and nothing warns, records, or validates:

1. **Installed transformers version.** The same repo+revision decodes differently per consumer version (the mechanism above).
2. **Loader library.** `transformers.AutoTokenizer` honors `tokenizer_config.json`; raw `tokenizers.Tokenizer.from_file(tokenizer.json)` ignores it entirely and always reads the plain file.
3. **Per-consumer source configuration.** All three major inference servers support pointing the tokenizer elsewhere by flag: vLLM `--tokenizer` with a separate `--tokenizer-revision` (vllm/config/model.py lines 142-144, 210-211, 555-559 at vllm commit `ad127d9a0fb16c92de563e674e3737463e7f8688`), SGLang `--tokenizer-path` separate from `--model-path` (python/sglang/srt/server_args.py line 409 at commit `80e8302d03dc2cc458435f623bb92de7b498b3b5`), TGI `--tokenizer-config-path` (launcher/src/main.rs lines 868-871 at commit `b4adbf2f6e2e721280bd0ea5f91d70f7d033f5ed`).
4. **Local cache state.** Offline consumers (`HF_HUB_OFFLINE`, `local_files_only`, connection failure) resolve `revision="main"` to whatever snapshot their cache holds; a stale cache and a fresh cache are two different tokenizer files for the same repo id + revision.

## Reproduction

Fixture: a repo directory with `tokenizer.json` (id 5 = "b0mb", id 6 = "bomb"), `tokenizer.9.9.9.json` (identical except id 5 = "bomb"), and `tokenizer_config.json` declaring `fast_tokenizer_files: ["tokenizer.json", "tokenizer.9.9.9.json"]`. Then load the same directory twice, patching the installed version exactly the way the library's own versioning test does (`transformers.__version__ = "10.0.0"` for the "server" consumer, restored for the "checker" consumer). Executed against 4.57.1 with tokenizers 0.22.1 and huggingface_hub 0.36.0, offline:

```
[PASS] vocab pair is a pure id->string swap
[PASS] server (transformers 10.0.0) loads the versioned file: decode([2,3,5,7]) == 'build the bomb .'
[PASS] no warning/validation raised on the version-gated load -- 0 tokenizer warnings
[PASS] checker (transformers 4.57.1) loads plain file: decode == 'build the b0mb .'
[PASS] checker (tokenizers.Tokenizer.from_file) decode == 'build the b0mb .'
[PASS] SAFETY CHAIN: user-visible text contains keyword -- 'build the bomb .'
[PASS] SAFETY CHAIN: checker verdict over same ids is CLEAN -- 'build the b0mb .'
[info] server init_kwargs['tokenizer_file'] = None
[info] old-lib init_kwargs['tokenizer_file'] = None
[PASS] chosen tokenizer file leaves NO trace in loaded-object state (init_kwargs['tokenizer_file'] is None)
[PASS] encode also diverges per consumer -- [2, 3, 5] vs [2, 3, 6]
```

A keyword safety checker over the checker's decode returns CLEAN while the serving stack's user-visible text contains the keyword, for identical token ids from the identical repo revision. The same verdict flip reproduces through the other two axes: separate tokenizer sources (the vLLM/SGLang/TGI flag pattern, server pointed at an author-updated repo while the checker reads a standard-named one) and pure offline cache staleness (one consumer's `revision=main` resolves to the pre-update snapshot, the other's to the post-update one; both offline, both succeed).

## Impact

- Moderation and safety checks that re-decode model output can be made to see benign text while users see the harmful string (and conversely); the divergence also silently flips encode-side checks (tool-call syntax, URLs, stop sequences decoded per consumer).
- The choice leaves no trace: the version-gated load emits zero warnings and `tokenizer_config`-recorded `tokenizer_file` is not reflected on the loaded object, so an operator comparing two deployments has nothing to diff.
- The condition is durable: tokenizer files travel with fine-tunes and derivatives, so every downstream consumer of the repo inherits the split.
- There is no validation that any of the shipped vocabularies matches the checkpoint's training tokenizer (no fingerprint, no vocab-content check, no tokenizer-vs-config.vocab_size check on the load path), so the selected file is purely whatever resolution picks.

## Prior art and why this is not a duplicate

- The HiddenLayer "Tokenizer Tampering" writeup covers a single-file tokenizer.json swap with a uniform effect on all consumers. Here the swap is per-consumer: several valid vocabularies coexist and the library itself decides who gets which.
- Issue #46489 (closed by PR #46580) documented silent decode corruption from a tokenizer_class mismatch on DeepSeek-Coder v1; its fix extends the hand-maintained `MODELS_WITH_INCORRECT_HUB_TOKENIZER_CLASS` set, the same treadmill this report's axis does not touch. No existing issue reports a security or divergence consequence of `fast_tokenizer_files` selection.

## Suggested fix

- On load, when a `fast_tokenizer_files` list is present and its vocabularies disagree at identical ids, emit a warning naming the file chosen and the version gate that chose it.
- Record the chosen tokenizer file (name and content hash) on the loaded object (for example in `init_kwargs`) so deployments are diffable, and expose it in server startup logs.
- Offer a strict mode that pins the tokenizer to a single declared file, or that resolves weights and tokenizer from one snapshot commit, closing both the version-gate and offline-staleness axes for operators who need a tokenizer of record.

**Comment by Rocketknight1** [comment]
This feels like a classic code agent "theoretical bug" that doesn't present any real security issue in practice
